### ISSUE TITLE
V9938/9958

### ISSUE THREAD

**Issue by cesss** [TomHarte/CLK#1100]
I didn't know about your emulator. Just found it today, and I really love your approach, which is very different from other emulators. Congrats and thanks a lot for writing it!

I've seen that you have written the 9918 emulation considering adding support to V9938/58 at some point. I didn't find any open issue for this, and these VDPs are my 8bit favourite gfx chips by far, so I just wished to say "don't forget these chips" 😃 

BTW, their DAC values are peculiar (the 58 differs from the 38, and, IIRC, the 2bit blue in SCREEN 8 is not evenly distributed).

Apart from that, I guess that their most time-consuming part at implementing them will be the VDP commands (blitter operations, lines, rects, color search)... they are pretty basic commands, but, it's more work than a plain framebuffer.

**Comment by TomHarte** [comment]
It is partly the accelerated drawing commands — primarily their timing — which has caused me a bit of a pause but in any case I think it'd be quite valuable to reopen this can of worms. After a false start a couple of months ago I think I want to perform a minor refactoring so as cleanly to be able to handle both the Yamaha and Sega descendants; likely the outer timing loop needs to exist separately from the inner drawing stuff.

Then the idea would be to implement the 9938/58 and the Mega Drive VDP, and hopefully add the MSX2 and Mega Drive as supported machines.

I already have the OPLL in there, so including the MSX-MUSIC should also be in scope; my OPLL was built with reconstitution of parts into an OPN in mind so the Sega side is not entirely unreachable.

So, if anything, the MSX2, and then the Mega Drive, should be this emulator's lowest-hanging fruits.

**Comment by cesss** [comment]
Yes, regarding the timing of the VDP commands, I took a look years ago at how other emulators do it, and I found just dumb delay hacks from seemingly experimental hand-made timing measures. I don't know if any site documents accurately the timing of commands.

**Comment by TomHarte** [comment]
I've started chipping away at it, at least. I guess that as long as I follow [Grauw's V9938 timing measurements](http://map.grauw.nl/articles/vdp-vram-timing/vdp-timing.html) and implement the commands as state machines that get only the documented access windows then I shouldn't be far off, and can certainly remedy problems in the future if and when they become clear.

In terms of my particular code as currently written, it's a hassle that the V9938 operates on a different internal clock from the TMS and the Master System VDP, but the Mega Drive operates on a different clock again so just breaking that assumption is justified.

My current implementation is slightly more than five years old so there's also going to be a brief period of refamiliarisation but it shouldn't be too bad. We'll see. The V9938 is quite a bit more advanced than the TMS so it might take a short while.

**Comment by cesss** [comment]
Grauw's info is great. I followed his list of DAC values for displaying MSX2/2+ images on current Macs considering the different gamma in TV CRTs and in current LCDs, and the result was great.
